### PR TITLE
Feat/2940 improve gesture handling inside image preview activity

### PR DESCRIPTION
**Pull request by TGazica**
### 🎯 Goal

Improve gesture handling inside `ImagePreviewActivity` by adding pan.

### 🛠 Implementation details

Add pan to `ImagePreviewActivity`.

### 🎨 UI Changes

<table>
<thead>
<tr>
<th>Before</th>
<th>After</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<video src="https://user-images.githubusercontent.com/38032787/162993846-32888fe3-c330-46bf-86c9-bcd737f33452.mp4" controls="controls" muted="muted" />
</td>
<td>
<video src="https://user-images.githubusercontent.com/38032787/162993937-69117fdd-5a40-4bcd-a5ce-7220fb1eb3f6.mp4" controls="controls" muted="muted" />
</td>
</tr>
</tbody>
</table>

### 🧪 Testing

Open an image, try to zoom, pan and swipe through galery.

#### General
- [x] Assigned a person / code owner group (required)
- [x] Thread with the PR link started in a respective Slack channel (#android-chat-core or #android-chat-ui) (required)
- [x] PR targets the `develop` branch
- [x] PR is linked to the GitHub issue it resolves

#### Code & documentation
- [x] Changelog is updated with client-facing changes
- [ ] New code is covered by unit tests
- [x] Comparison screenshots added for visual changes
- [x] Affected documentation updated (KDocs, docusaurus, tutorial)

### ☑️Reviewer Checklist
- [ ] UI Components sample runs & works
- [ ] Compose sample runs & works
- [ ] UI Changes correct (before & after images)
- [ ] Bugs validated (bugfixes)
- [ ] New feature tested and works
- [ ] Release notes and docs clearly describe changes
- [ ] All code we touched has new or updated KDocs

### 🎉 GIF

giphy

### CODE REVIEW

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/attachments/preview/ImagePreviewActivity.kt [hunk 1]
```diff
@@ -475,52 +494,136 @@ public class ImagePreviewActivity : AppCompatActivity() {
...
+                var currentScale by remember { mutableStateOf(1f) }
+                var translation by remember { mutableStateOf(Offset(0f, 0f)) }
+
+                val scale by animateFloatAsState(targetValue = currentScale)
+
+                Image(
+                    modifier = Modifier
+                        .then(
+                            (painter.state as? ImagePainter.State.Success)
+                                ?.painter
```
Comment (filbabic) [inline stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/attachments/preview/ImagePreviewActivity.kt:516]: I would pull all of this out into another value:

```
                val transformModifier = if (painter.state is ImagePainter.State.Success) {
                    val size = painter.intrinsicSize

                    Modifier.aspectRatio(size.width / size.height, true)
                } else {
                    Modifier
                }
```

Then apply it like so:

```
transformModifier
.graphicsLayer()
....
```

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/attachments/preview/ImagePreviewActivity.kt [hunk 2]
```diff
@@ -475,52 +494,136 @@ public class ImagePreviewActivity : AppCompatActivity() {
...
+                                        val event = awaitPointerEvent(pass = PointerEventPass.Initial)
+
+                                        val maxTranslation = calculateMaxOffset(
+                                            imageSize = imageSize,
+                                            scale = currentScale,
+                                            parentSize = parentSize
+                                        )
+
+                                        val zoom = event.calculateZoom()
+                                        currentScale = (zoom * currentScale).coerceAtMost(3f)
```
Comment (filbabic) [inline stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/attachments/preview/ImagePreviewActivity.kt:545]: should pull out as a constant so it's not a magic number (e.g. `DefaultMaxImageZoomScale`)

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/attachments/preview/ImagePreviewActivity.kt [hunk 3]
```diff
@@ -475,52 +494,136 @@ public class ImagePreviewActivity : AppCompatActivity() {
...
+                                        translation = Offset(
+                                            newTranslationX.coerceIn(-maxTranslation.x, maxTranslation.x),
+                                            newTranslationY.coerceIn(-maxTranslation.y, maxTranslation.y)
+                                        )
+
+                                        if (abs(newTranslationX) < calculateMaxOffsetPerAxis(
+                                                imageSize.width,
+                                                currentScale,
+                                                parentSize.width
+                                            ) || zoom != 1f
```
Comment (filbabic) [inline stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/attachments/preview/ImagePreviewActivity.kt:560]: These out as well (`DefaultZoomScale`)

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/attachments/preview/ImagePreviewActivity.kt [hunk 4]
```diff
@@ -475,52 +494,136 @@ public class ImagePreviewActivity : AppCompatActivity() {
...
-            )
+                        }
+                        .pointerInput(Unit) {
+                            forEachGesture {
+                                awaitPointerEventScope {
+                                    awaitFirstDown()
+                                    withTimeoutOrNull(DOUBLE_TAP_TIMEOUT_MS) {
+                                        awaitFirstDown()
+                                        currentScale = when {
+                                            currentScale == 3f -> 1f
```
Comment (filbabic) [inline stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/attachments/preview/ImagePreviewActivity.kt:579]: `MaxZoomScale`

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/attachments/preview/ImagePreviewActivity.kt [hunk 5]
```diff
@@ -475,52 +494,136 @@ public class ImagePreviewActivity : AppCompatActivity() {
...
+                                            currentScale >= 2f -> 3f
```
Comment (filbabic) [inline stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/attachments/preview/ImagePreviewActivity.kt:580]: `MidZoomScale` or something in that sense.

**Comment by TGazica** [comment]
> 

The changes above have been implemented.

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/attachments/preview/ImagePreviewActivity.kt [hunk 6]
```diff
@@ -475,52 +494,136 @@ public class ImagePreviewActivity : AppCompatActivity() {
...
-                    .transformable(state = transformableState)
-                    .pointerInput(Unit) {
-                        detectTapGestures(
-                            onDoubleTap = {
-                                val newScale = when {
-                                    currentScale == 3f -> 1f
-                                    currentScale >= 2f -> 3f
-                                    else -> 2f
+                Image(
+                    modifier = Modifier
```
Comment (filbabic) [inline stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/attachments/preview/ImagePreviewActivity.kt:520]: can remove the base modifier and just use transformModifier as the base

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/attachments/preview/ImagePreviewActivity.kt [hunk 7]
```diff
@@ -475,52 +494,136 @@ public class ImagePreviewActivity : AppCompatActivity() {
...
+                            }
+                        },
+                    painter = painter,
+                    contentDescription = null
+                )
 
-            if (pagerState.currentPage != page) {
-                currentScale = 1f
+                if (pagerState.currentPage != page) {
+                    currentScale = 1f
```
Comment (filbabic) [inline stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/attachments/preview/ImagePreviewActivity.kt:596]: can be a cosntant here I think

**Review by filbabic — APPROVED** [review]

**Comment by filbabic** [comment]
- [x] Compose sample runs & works
- [x] UI Changes correct (before & after images)
- [x] New feature tested and works
- [x] Release notes and docs clearly describe changes
- [x] All code we touched has new or updated KDocs

File: stream-chat-android-offline/detekt-baseline.xml [hunk 8]
```diff
@@ -165,7 +165,6 @@
     <ID>MaxLineLength:UploadAttachmentsNetworkType.kt$UploadAttachmentsNetworkType$* An enumeration of various network types used as a constraint in [io.getstream.chat.android.offline.internal.message.attachments.UploadAttachmentsAndroidWorker].</ID>
     <ID>MaxLineLength:UserEntity.kt$UserEntity$*</ID>
     <ID>MaxLineLength:WhenEnrichChannel.kt$WhenEnrichChannel$fun</ID>
-    <ID>MaxLineLength:WhenUpdateLastMessage.kt$WhenUpdateLastMessage$fun</ID>
```
Comment (filbabic) [inline stream-chat-android-offline/detekt-baseline.xml:168]: any idea why this was dropped?

Comment (TGazica) [reply stream-chat-android-offline/detekt-baseline.xml:168]: No, i ran detektBaseline, could be that the issue was resolved but detektBaseline wasn't run?

Comment (filbabic) [reply stream-chat-android-offline/detekt-baseline.xml:168]: yup most likely!

**Pull request merged**